pkg/plan/track_response_builder.go

```diff
@@ -0,0 +1,216 @@
+		id, refID = stringPFieldValue(v, "ID"), stringPFieldValue(v, "RefID")
+	}
+
+	step, err := GetStepName(stepLog)
+	if step == "" {
+		return TrackResponse{}, ErrPlanFinished
+	}
+
+	return TrackResponse{
+		Kind:     kind,
+		ID:       id,
+		RefID:    refID,
+		Step:     step,
+		Err:      err,
+		Finished: step == planCompleted,
+		Duration: getPlanDuration(stepLog),
+	}, nil
+}
+
+// getPlanStepInfo takes in a resource's info and returns a slice of
+// ClusterPlanStepInfo, which is obtained by accessing the named fields via
+// reflection. The flow might be confusing but in a nutshell it tries to:
+//   1. Obtain the Pending plan step log.
+//   2. Obtain the Current plan step log when getCurrentPlan is true.
+//   3. (if getCurrentPlan == true and the Current plan is empty) obtains the
+//       "Current" plan accessing the last item in the plan history slice.
+func getPlanStepInfo(wl interface{}, getCurrentPlan bool) ([]*models.ClusterPlanStepInfo, error) {
+	var planName = pendingPlanPath
+	if getCurrentPlan {
+		planName = currentPlanPath
+	}
+
+	payloadValue := reflect.ValueOf(wl)
+	if !payloadValue.IsValid() {
+		return nil, errNoPendingPlan
+	}
+
+	// Get either the "Pending" or "Current" plan.
+	plan := reflectElemFieldPath(payloadValue, planName)
+	if !plan.IsValid() {
+		return nil, errNoPendingPlan
+	}
+
+	// If the pending plan is nil and getCurrentPlan == false, return an error.
+	var noPendingPlanAndWantPendingPlan = plan.IsNil() && !getCurrentPlan
+	if noPendingPlanAndWantPendingPlan {
+		return nil, errNoPendingPlan
+	}
+
+	// When either "Pending" or "Current" aren't nil, obtain the plan log.
+	var planLog reflect.Value
+	if !plan.IsNil() {
+		planLog = plan.Elem().FieldByName(planAttemptLog)
+	}
+
+	// When either "Pending" or "Current" are nil, and the current plan needs to
+	var currentPlanIsNilAndPlanLogIsNil = plan.IsNil() && getCurrentPlan || planLog.IsNil()
+	if currentPlanIsNilAndPlanLogIsNil {
+		if history := reflectElemFieldPath(payloadValue, historyPlanPath); history.Len() > 0 {
+			var lastPlan = history.Index(history.Len() - 1)
+			planLog = lastPlan.Elem().FieldByName(planAttemptLog)
+		}
+	}
+
+	return getPlanLog(planLog)
+}
+
+func getPlanDuration(log []*models.ClusterPlanStepInfo) strfmt.Duration {
+	for i := range log {
+		return strfmt.Duration(time.Since(time.Time(*log[i].Started)))
+	}
+
+	return 0
+}
+
+// stringPFieldValue obtains the value of a string pointer field.
+func stringPFieldValue(v reflect.Value, field string) string {
+	if value := v.Elem().FieldByName(field); value.IsValid() {
+		valueStrP := value.Interface().(*string)
+		return *valueStrP
+	}
+	return ""
+}
+
+// reflectElemFieldPath obtains the reflect.Value at the end of the specified
+// path. Path is in the format of <Property>.<Property> as many times as
+// required to obtain the end field.
+func reflectElemFieldPath(v reflect.Value, p string) reflect.Value {
+	props := strings.Split(p, ".")
```

Comment:
I know it's a lot longer, but I think writing out `properties` might make sense in this case, because `props` can mean two other things, and it's a little confusing.